server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java

```diff
@@ -248,14 +249,47 @@ protected AllocateUnassignedDecision getAllocationDecision(
         final RoutingNodes routingNodes = allocation.routingNodes();
         final boolean explain = allocation.debugDecision();
         ShardRouting primaryShard = routingNodes.activePrimary(unassignedShard.shardId());
+
         if (primaryShard == null) {
-            assert explain : "primary should only be null here if we are in explain mode, so we didn't "
-                + "exit early when canBeAllocatedToAtLeastOneNode didn't return a YES decision";
-            return AllocateUnassignedDecision.no(
-                UnassignedInfo.AllocationStatus.fromDecision(allocationDecision.v1().type()),
-                new ArrayList<>(allocationDecision.v2().values())
-            );
+            // Determine if the index is configured for search-only.
+
+            boolean isIndexSearchOnly = allocation.metadata()
+                .getIndexSafe(unassignedShard.index())
+                .getSettings()
+                .getAsBoolean(IndexMetadata.INDEX_BLOCKS_SEARCH_ONLY_SETTING.getKey(), false);
+
+            // Only proceed if both the shard is marked search-only and the index setting is enabled.
+            if (unassignedShard.isSearchOnly() && isIndexSearchOnly) {
```

Comment:
Lets look up the index setting only if the shard is searchOnly? - 

```java
if (unassignedShard.isSearchOnly) {
  return getSearchReplicaAllocationDecision(....);
}
```

Reply:
Let me created `getSearchReplicaAllocationDecision` and move the logic there and also move the `isIndexSearchOnly` inside the `if (unassignedShard.isSearchOnly()`.